### DISCUSSION TITLE
useState Strokes average

### DISCUSSION THREAD

**Discussion by fimoefive** [nss-evening-cohort-14/evening-client-side#320] [Capstones!] [answered]
Martin

**issue:** Having trouble trying to add up all the values or different key: names. example:
(hole1 + hole2 + hole3 + hole4 + hole5 + hole6 + hole7 + hole8 + hole9)
const [average, setAverage] = useState([]);

**What I have tried**:
I have a .forEach function and .reduce function in my views.js file. 
Also I tried a handler in the form file for an onChange event for the input fields.
I believe the easiest way to go about this is to loop through all the key: 'values' .forEach card and have a function some and divide by .length.
 
I've also setup my Firebase rules to:
``` js 

"holes": {
        ".indexOn": ["hole1", "hole2", "hole3", "hole4", "hole5", "hole6", "hole7", "hole8", "hole9", ".value"]

  // const valuesHandler = (e) => {
  //   const { name } = e.target;
  //   const { value } = e.target.value;
  //   const newValues = {
  //     ...hole,
  //     [name]: value
  //   };
  //   setHole(newValues);
  //   calcTotal(newValues);
  // };

 const holeCal = () => {
    let holeTotal = 0;
    const hLength = holes.length;
    holes.forEach(({ hole }) => holeTotal += hole);
    return holeTotal / hLength;
  };

  function avg() {
    let hole = 0;
    holes.forEach((obj) => {
      hole += obj.hole;
    });
    hole /= holes.length;
    return hole;
  }
```
**Google links:**
https://stackoverflow.com/questions/43260689/adding-and-indexon-value-into-firebase-rules
https://jrsinclair.com/articles/2019/five-ways-to-average-with-js-reduce/
https://stackoverflow.com/questions/61349874/how-to-sum-values-of-multiple-inputs-using-onchange-event-reactjs
https://stackoverflow.com/questions/62358365/react-js-get-sum-of-numbers-in-array

**Links to code:**
https://github.com/fimoefive/Golf-ScoreCard-App/blob/main/src/views/Holes.js
https://github.com/fimoefive/Golf-ScoreCard-App/blob/main/src/forms/HoleForm.js

**Comment by ajawashington** [comment, marked as answer]
@fimoefive Where are you placing this functionality in your code? I don't see it in any of your links here.

**Reply by fimoefive** [reply]
I am placing it in Holes.js the links are my previous branch and not my current working branch. 
Should I have a .filter for each key name for my card that I'm trying to add up; since the key: names are not uniformly the same? I'm trying to add each hole value and divide by the length. 
I can push up my current working branch if that helps.

**Reply by ajawashington** [reply]
Also on this line https://github.com/fimoefive/Golf-ScoreCard-App/blob/29305b1fbeae6fba22e532ade4d6a927a72b1d78/src/helpers/data/holeData.js#L30 
You are passing it a user but the `getHoles` call does not have any parameters.

**Reply by ajawashington** [reply]
@fimoefive could you push up your current working branch and link it.

**Reply by fimoefive** [reply]
[average working branch](https://github.com/fimoefive/Golf-ScoreCard-App/tree/mjs-playerAvg)

I went off in the wrong direction with these dynamic functions. I need help getting back on track and trying to determine the simplest way to calculate the sum/average of the players strokes. I've spent all day yesterday on this and feel like I'm just spinning my wheels.
Thanks

**Reply by fimoefive** [reply]
After writing various for loops Trinity helped me add all the holes by restructuring each hole and was able to add them using Number method instead of parsInt method to get the game stroke average. 

  function avg(holeInfo) {
    // console.warn(holeInfo);
    let hole = 0;
    hole += Number(holeInfo.hole1);
    hole += Number(holeInfo.hole2);
    hole += Number(holeInfo.hole3);
    hole += Number(holeInfo.hole4);
    hole += Number(holeInfo.hole5);
    hole += Number(holeInfo.hole6);
    hole += Number(holeInfo.hole7);
    hole += Number(holeInfo.hole8);
    hole += Number(holeInfo.hole9);
    // console.warn(hole);
    hole /= 9;
    console.warn(holeInfo[`hole${2}`]);
    return hole;
  }
